File: basic_dsa_one_hundred_fifty_six.java
/**
 * Basic DSA Problems - Climbing Stairs
 * Number of distinct ways to climb n steps (1 or 2 steps at a time)
 */
public class basic_dsa_one_hundred_fifty_six {

    // Problem: Climb n steps. Each time you can climb 1 or 2 steps. Return number of distinct ways.

    /**
     * dp[i] = ways to reach step i. dp[i] = dp[i-1] + dp[i-2]. Same as Fibonacci.
     * Time O(n), Space O(1)
     */
    public static int climbStairs(int n) {
        if (n <= 2) return n;
        int a = 1, b = 2;
        for (int i = 3; i <= n; i++) {
            int next = a + b;
            a = b;
            b = next;
        }
        return b;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Climbing Stairs");
        System.out.println("=".repeat(60));

        System.out.println("2 -> " + climbStairs(2));
        System.out.println("3 -> " + climbStairs(3));
        System.out.println("5 -> " + climbStairs(5));
    }
}

File: basic_dsa_one_hundred_fifty_six.py
"""
Basic DSA - Climbing Stairs
You are climbing a staircase. It takes n steps to reach the top.
Each time you can climb 1 or 2 steps. Return the number of distinct ways to climb.
"""


def climb_stairs(n):
    """
    Fibonacci DP with O(1) space.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if n <= 2:
        return n
    a, b = 1, 2
    for _ in range(3, n + 1):
        a, b = b, a + b
    return b


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Climbing Stairs")
    print("=" * 60)

    print(climb_stairs(2))  # 2
    print(climb_stairs(3))  # 3
    print(climb_stairs(5))  # 8
